runtime/compiler/runtime/Listener.cpp

```diff
@@ -20,18 +20,284 @@
+
+   bio = (*OBIO_new_ssl)(sslCtx, false);
+   if (!bio)
+      return handleOpenSSLConnectionError(connfd, ssl, bio, "Error creating new BIO");
+
+   if ((*OBIO_ctrl)(bio, BIO_C_SET_SSL, true, (char *)ssl) != 1) // BIO_set_ssl(bio, ssl, true)
+      return handleOpenSSLConnectionError(connfd, ssl, bio, "Error setting BIO SSL");
+
+   if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+      TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "SSL connection on socket 0x%x, Version: %s, Cipher: %s\n",
+                                                     connfd, (*OSSL_get_version)(ssl), (*OSSL_get_cipher)(ssl));
+   return true;
+   }
 
 TR_Listener::TR_Listener()
    : _listenerThread(NULL), _listenerMonitor(NULL), _listenerOSThread(NULL), 
    _listenerThreadAttachAttempted(false), _listenerThreadExitFlag(false)
    {
    }
 
+void
+TR_Listener::serveRemoteCompilationRequests(BaseCompileDispatcher *compiler)
+   {
+   TR::PersistentInfo *info = getCompilationInfo(jitConfig)->getPersistentInfo();
+   SSL_CTX *sslCtx = NULL;
+   if (JITServer::CommunicationStream::useSSL())
+      {
+      JITServer::CommunicationStream::initSSL();
+      sslCtx = createSSLContext(info);
+      }
+
+   uint32_t port = info->getJITServerPort();
+   uint32_t timeoutMs = info->getSocketTimeout();
+   struct pollfd pfd = {0};
+   int sockfd = socket(AF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
+   if (sockfd < 0)
+      {
+      perror("can't open server socket");
+      exit(1);
+      }
+
+   // see `man 7 socket` for option explanations
+   int flag = true;
+   if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, (void *)&flag, sizeof(flag)) < 0)
+      {
+      perror("Can't set SO_REUSEADDR");
+      exit(-1);
+      }
+   if (setsockopt(sockfd, SOL_SOCKET, SO_KEEPALIVE, (void *)&flag, sizeof(flag)) < 0)
+      {
+      perror("Can't set SO_KEEPALIVE");
+      exit(-1);
+      }
+
+   struct sockaddr_in serv_addr;
+   memset((char *)&serv_addr, 0, sizeof(serv_addr));
+   serv_addr.sin_family = AF_INET;
+   serv_addr.sin_addr.s_addr = htonl(INADDR_ANY);
+   serv_addr.sin_port = htons(port);
+
+   if (bind(sockfd, (struct sockaddr *)&serv_addr, sizeof(serv_addr)) < 0)
+      {
+      perror("can't bind server address");
+      exit(1);
+      }
+   if (listen(sockfd, SOMAXCONN) < 0)
+      {
+      perror("listen failed");
+      exit(1);
+      }
+
+   pfd.fd = sockfd;
+   pfd.events = POLLIN;
+
+   while (!getListenerThreadExitFlag())
+      {
+      int32_t rc = 0;
+      struct sockaddr_in cli_addr;
+      socklen_t clilen = sizeof(cli_addr);
+
+      rc = poll(&pfd, 1, OPENJ9_LISTENER_POLL_TIMEOUT);
+      if ((rc == 0) && getListenerThreadExitFlag())
+         {
+         break;
+         }
+      else if (rc < 0)
+         {
+         if (TR::Options::getVerboseOption(TR_VerboseJITServer))
+            TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer, "Error in polling socket: errno=%d", errno);
+	 exit(1);
+         }
```

Comment:
Do we have to check the flags set in `revent`? We could have `POLLIN`/`POLLERR`/`POLLHUP`/`POLLNVAL`

Reply:
We should never get `POLLNVAL` as it bad `fd`.
 `POLLHUP` can happen for connected sockets/pipes where one ends gets closed and we try to read from the other end. Again this does not apply to our case.
`POLLERR` indicates "an asynchronous error occurred" (http://man7.org/linux/man-pages/man7/socket.7.html). Doesn't look like it applies to listening socket.

Only flag left is `POLLIN` which indicates request for new connection.

Reply:
I see in this example: `https://www.ibm.com/support/knowledgecenter/ssw_ibm_i_71/rzab6/poll.htm` that a check of
```
      if(fds[i].revents != POLLIN)
      {
        printf("  Error! revents = %d\n", fds[i].revents);
        end_server = TRUE;
        break;
      }
```
is done. Maybe we should do the same to be sure.

Reply:
I have added a new commit with this change. I also added a do-while loop for accepting new connections as per the example mentioned in the link in previous comment, so that we can avoid calling `poll()` if multiple clients are ready.